e3sm_diags/driver/polar_driver.py

```diff
@@ -1,264 +1,400 @@
+            ds_test_region,
+            ds_test_region_regrid,
+            ds_ref_region,
+            ds_ref_region_regrid,
+            ds_diff_region,
         )
 
-        # Get land/ocean fraction for masking.
-        try:
-            land_frac = test_data.get_climo_variable("LANDFRAC", season)
-            ocean_frac = test_data.get_climo_variable("OCNFRAC", season)
-        except Exception:
-            mask_path = os.path.join(
-                e3sm_diags.INSTALL_PATH, "acme_ne30_ocean_land_mask.nc"
+        _save_data_metrics_and_plots(
+            parameter,
+            plot_func,
+            var_key,
+            ds_test_region,
+            ds_ref_region,
+            ds_diff_region,
+            metrics_dict,
+        )
+    
+def _run_diags_3d(
+    parameter: CoreParameter,
+    ds_test: xr.Dataset,
+    ds_ref: xr.Dataset,
+    ds_land_sea_mask: xr.Dataset,
+    season: str,
+    regions: List[str],
+    var_key: str,
+    ref_name: str,
+):
+    """Run diagnostics on a 3D variable.
+
+    This function gets the variable's metrics by region, then saves the
+    metrics, metric plots, and data (optional, `CoreParameter.save_netcdf`).
+
+    Parameters
+    ----------
+    parameter : CoreParameter
+        The parameter object.
+    ds_test : xr.Dataset
+        The dataset containing the test variable.
+    ds_ref : xr.Dataset
+        The dataset containing the ref variable. If this is a model-only run
+        then it will be the same dataset as ``ds_test``.
+    ds_land_sea_mask : xr.Dataset
+        The land sea mask dataset, which is only used for masking if the region
+        is "land" or "ocean".
+    season : str
+        The season.
+    regions : List[str]
+        The list of regions.
+    var_key : str
+        The key of the variable.
+    ref_name : str
+        The reference name.
+    """
+    ### NOTE: Here below is copied from lat_lon_driver.py
+    plev = parameter.plevs
+    logger.info("Selected pressure level(s): {}".format(plev))
+
+    ds_test_rg = regrid_z_axis_to_plevs(ds_test, var_key, parameter.plevs)
+    ds_ref_rg = regrid_z_axis_to_plevs(ds_ref, var_key, parameter.plevs)
+
+    for ilev in plev:
+        z_axis_key = get_z_axis(ds_test_rg[var_key]).name
+        ds_test_ilev = ds_test_rg.sel({z_axis_key: ilev})
+        ds_ref_ilev = ds_ref_rg.sel({z_axis_key: ilev})
+
+        for region in regions:
+            (
+                ds_test_region,
+                ds_test_region_regrid,
+                ds_ref_region,
+                ds_ref_region_regrid,
+                ds_diff_region,
+            ) = subset_and_align_datasets(
+                parameter,
+                ds_test_ilev,
+                ds_ref_ilev,
+                ds_land_sea_mask,
+                var_key,
+                region,
             )
-            with cdms2.open(mask_path) as f:
-                land_frac = f("LANDFRAC")
-                ocean_frac = f("OCNFRAC")
-
-        for var in variables:
```

Comment:
Again, the refactoring switches the loops for seasons and vars -- i.e., in the refactoring, vars is the outer loop.